0.7 (26 Jun 2018)

- Dockerized version with --docker
- Custom reference data with --pon, --ref-fasta, --truth-regions, or --bcbio-genomes

0.6 (15 May 2018)

- Use snakemake groups to optimize execution on cluster
- Use submission wrapper to automate cluster resources
- Filter structural variants: BPI_AF>=10% and read support>=5
- Keep sv-prioritize hard-filtered variants
- 0.6.1
  - Add circos plots into the Rmd report. Generated by https://github.com/umccr/rock

0.5 (7 May 2018)

- PCGR is deployed directly on Spartan, so no AWS dependency.
  - Add pcgr wrapper: `pcgr variants.vcf.gz cnv.tsv -o results [-g hg38]`
- Correctly providing memory resources on HPC to avoid oom-kill
- On Spartan, support `--cluster-auto` to automatically substitute proper cluster parameters

0.4 (17 Apr 2018)

- Propagate snakemake's cluster options to the wrapper
- Propagate snakemake's `--rerun-incomplete` and to the wrapper
- PCGR: stop using the UID, skip uploading if already uploaded
- Fix `HTTP_PROXY` setting for running on Spartan worker nodes
- New panel of normals: add the large A5 cohort, normalize normals, check only POS for indels
- Correctly set threads to mibibams samtools runs

0.3 (17 Jan 2018)

- Refactor output folder structure (see docs)
- `pcgr_download` taget to automatically pull results
- Automatically attempt downloading PCGR results after minibams
- Minibams always run in the ned
- Use `--unlock` to automate the snakemake unlock pattern to continue unterrupted runs
- Move `vcf_stuff` to a separate repo
- Remove GL* chromosomes from the cnvkit diagram
- PCGR tomls: changed mutsignatures_normalization to genome
- Rmd: add SV-prioritize table
