---
title: "UMCCR Patient Summary"
author: "Oliver Hofmann"
date: "`r Sys.Date()`"
output:
  html_document:
    theme: readable
    toc: false
    toc_float: false
    code_folding: hide
  rmdformats::material:
    highlight: kate
params:
  workdir: '/Users/vsaveliev/tmp'
  vcf_fname: 'ensemble-hg19.vcf'
  af_freqs: 'af/af_tumor.txt'
  af_freqs_az300: 'af/af_tumor_az300.txt'
  sv_fname: 'sv-prioritize-manta-bpi.tsv'
  tumor_name: '17MHP031Tmr'
  sig_probs: '/Users/vsaveliev/git/umccr/umccrise/umccrise/patient_analysis/rmd_files/signatures_probabilities.txt'
  suppressors: '/Users/vsaveliev/git/umccr/umccrise/umccrise/patient_analysis/rmd_files/suppressors.txt'
---

```{r custom, echo=FALSE, message=FALSE, warning=FALSE}
library(ggplot2)
library(knitr)
library(rmarkdown)
library(tidyr)
library(dplyr)
library(readr)
library(DT)
library(MutationalPatterns)
library(BSgenome)
library(devtools)
library(stringr)
ref_genome <- 'BSgenome.Hsapiens.UCSC.hg19'
library(ref_genome, character.only = TRUE)
```

```{r render_interactively, eval=FALSE, echo=FALSE}
renderMyDocument <- function() {
  rmarkdown::render("sig.Rmd", params = list(
    tumor_name = 'sample_foo',
    vcf_fname = '../../nogit/data/somatic/ensemble-hg19.vcf',
    af_freqs = '../../nogit/data/af/af_vardict_tumor.txt',
    af_freqs_az300 = '../../nogit/data/af/af_vardict_tumor_az300.txt',
    sv_fname = '../../nogit/data/structural/sv-prioritize-manta-bpi.tsv',
    sig_probs = 'rmd_files/signatures_probabilities.txt',
    supressors = 'rmd_files/suppressors.txt'
  ))
}

# can run this interactively to knit report
renderMyDocument()
```

```{r "setup", include=FALSE}
knitr::opts_knit$set(root.dir = params$workdir)
options(width = 140)
```


## Allelic frequencies

Comparing the allelic frequencies of all somatic mutations vs those present in a subset of ~300 known cancer genes. Frequencies are currently based on Mutect2 calls only and limited to 'high confidence' regions as determined by the [Genome in a Botte consortium].

```{r af}
# Global AF
af <- readr::read_tsv(params$af_freqs, col_names = "af", col_types = "d") %>%
  dplyr::mutate(set = "wgs")

# AF for AstraZeneca's AZ300 gene set
af_az300 <- readr::read_tsv(params$af_freqs_az300, col_names = "af", col_types = "-----d-") %>%
  dplyr::mutate(set = 'cancer_genes')

dplyr::bind_rows(af, af_az300) %>%
  ggplot(aes(af)) +
  geom_histogram(stat = 'bin', binwidth = 0.01, fill = "firebrick") +
  facet_wrap(~set, scales = 'free_y') +
  theme_bw()
```

```{r modecalc}
mode <- function(x) {
  ux <- unique(x)
  ux[which.max(tabulate(match(x, ux)))]
}
```
Based on all variants we are seeing a mean allelic frequency of `r round(mean(af$af), 2)` (median `r median(af$af)`, mode `r mode(af$af)`). Looking at just SNVs in known cancer genes this changes to mean `r round(mean(af_az300$af), 2)` (median `r median(af_az300$af)`, mode `r mode(af_az300$af)`).


## Mutational signature

[PCGR](https://github.com/sigven/pcgr) uses [deconstructSigs](https://cran.r-project.org/web/packages/deconstructSigs/index.html) to generate somatic signatures. From our experience it can miss signatures quite frequently; using [MutationalPatterns](http://bioconductor.org/packages/release/bioc/html/MutationalPatterns.html) to doublecheck.

```{r importVCF}
# Grab ensemble calls from the post-processing stage. These were converted
# by adding "chr" prefixes and remiving G* chromosomes.
#

vcf <- read_vcfs_as_granges(
  params$vcf_fname,
  sample_names = params$tumor_name,
  genome = "BSgenome.Hsapiens.UCSC.hg19",
  group = "auto+sex")
```

### Look at somatic change distribution:

```{r somProfile}
type_occurrences <- mut_type_occurrences(vcf, ref_genome = ref_genome)
plot_spectrum(type_occurrences, CT = TRUE)

# Could do this relative to a few reference samples
mut_mat <- mut_matrix(vcf_list = vcf, ref_genome = ref_genome)
plot_96_profile(mut_mat)
```

### Extract somatic signatures and compare to the reference set:

```{r somSig}
# Get Sanger sigs
# sp_url <- paste("http://cancer.sanger.ac.uk/cancergenome/assets/", "signatures_probabilities.txt", sep = "")
sp_url <- params$sig_probs
cancer_signatures <- readr::read_tsv(sp_url, col_names = TRUE) %>%
  dplyr::arrange(`Substitution Type`) %>%
  dplyr::select(4:33) %>%
  as.matrix()

# Fit mutation matrix to cancer signatures. This function finds the optimal linear
# combination of mutation signatures that most closely reconstructs the
# mutation matrix by solving a non-negative least-squares constraints problem.
fit_res <- fit_to_signatures(mut_mat, cancer_signatures)
# Select signatures with some contribution
fit_res_contr <- fit_res$contribution[fit_res$contribution[, 1] > 0, ]
result <- data.frame(Signature = names(fit_res_contr), Contribution = fit_res_contr)

# Quick summary of the results; no plotting yet
result %>%
  dplyr::select(Signature, Contribution) %>%
  dplyr::arrange(-Contribution) %>%
  DT::datatable(rownames = FALSE) %>%
  DT::formatRound('Contribution', 1)
```

### Check for positional enrichment of somatic signatures (limited to autosomes):
```{r rainfall}
chromosomes <- seqnames(get(ref_genome))[1:22]
plot_rainfall(vcf[[1]], chromosomes = chromosomes, cex = 1.5, ylim = 1e+09 )
```


## Prioritised structural variants

```{r sv_prioritize}
cnv_all <- read.table(params$sv_fname,
                      header= TRUE, sep="\t", stringsAsFactors = FALSE)

# SV events
cnv <- cnv_all %>%
  tidyr::unnest(annotation = strsplit(annotation, ',')) %>% # Unpack multiple annotations per region
  tidyr::separate(annotation,
           c('type', 'annotation', 'gene', 'transcript', 'priority', 'score'),
           sep='\\|') %>% # Unpack annotation columns
  dplyr::select(chrom, start, end, gene, priority, score, annotation, type,
                split_read_support, paired_support_PR) %>%
  dplyr::distinct()
datatable(cnv, rownames=FALSE)
```

## Loss of function in tumor suppressors

```{r lof}
suppressors <- read.table(params$suppressors, header=FALSE)$V1

lof <- cnv_all %>%
  mutate(lof=str_replace_all(lof, '\\(|\\)', '')) %>%
  tidyr::unnest(lof = strsplit(lof, ',')) %>%
  tidyr::separate(lof,
           c('lof_gene', 'geneid', 'transcript_count', 'percent_affected'),
           sep='\\|') %>%
  dplyr::select(chrom, start, end, lof_gene, transcript_count, percent_affected,
         split_read_support, paired_support_PR) %>%
  dplyr::filter(lof_gene %in% suppressors) %>%
  dplyr::distinct()
datatable(lof, rownames=FALSE)
```

## Addendum

```{r sessioninfo}
devtools::session_info()
```

